exercise-book/src/simple-db-solution.md

```diff
@@ -0,0 +1,204 @@
+
+<details>
+  <summary>Solution</summary>
+
+```rust, ignore
+{{#include ../../exercise-solutions/simple_db/step4b/src/lib.rs:18:38}}
+```
+
+</details>
+
+### Step 4c: Pattern matching for the command keywords
+
+Remove the Ok(Command::Command) and the enum variant. Use `.trim()` on the command substring and use `match` to patternmatch its content. `.trim()` removes any `\n` that are in the substring. Next, implement two necessary match arms: `""` for empty messages, `_` for any other string, currently evaluated to be an unknown command.
+
+<details>
+  <summary>Solution</summary>
+
+```rust, ignore
+{{#include ../../exercise-solutions/simple_db/step4c/src/lib.rs:17:39}}
+```
+
+</details>
+
+### Step 4d: Add Retrieve Case
+
+Add a match arm to check if the command substring is equal to `"RETRIEVE"`. It’s not enough to return `Ok(Command::Retrieve)` just yet. The Retrieve command cannot have a payload, this includes whitespace! To check for this, add an if else statement, that checks if the next iteration over the substrings returns none. If this is true, return the `Ok(Command::Retrieve)`, if it is false, return `Err(Error::UnexpectedPayload)`.
+
+<details>
+  <summary>Solution</summary>
+
+```rust, ignore
+{{#include ../../exercise-solutions/simple_db/step4d/src/lib.rs:17:46}}
+```
+
+</details>
+
+### Step 4e: Add Publish Case and finish
+
+Add a `match` arm to check if the command substring is equal to `"PUBLISH"`. Just like with the Retrieve command, we need to add a distinction, but the other way round: Publish needs a payload or whitespace for an empty payload to be valid.
+
+Use `if let` to check if the next iteration into the substrings returns `Some()`. If it does, return `Ok(Command::Publish(T))`, where T is an owned version of the trimmed payload. Otherwise return `Err(Error::MissingPayload)`.
+
+<details>
+  <summary>Solution</summary>
+
+```rust, ignore
+{{#include ../../exercise-solutions/simple_db/step4e/src/lib.rs:17:53}}
+```
+
+</details>
+
+## Full source code
+
+If all else fails, feel free to copy this solution to play around with it.
+
+<details>
+  <summary>Solution</summary>
+
+```rust
+    #[derive(Eq, PartialEq, Debug)]
+    pub enum Command {
+        Publish(String),
+        Retrieve,
+    }
+
+    #[derive(Eq, PartialEq, Debug)]
+    pub enum Error {
+        TrailingData,
+        IncompleteMessage,
+        EmptyMessage,
+        UnknownCommand,
+        UnknownError,
+        UnexpectedPayload,
+        MissingPayload,
+    }
+
+    pub fn parse(input: &str) -> Result<Command, Error> {
+        match input.split_once('\n') {
+            Some((_, data)) => {
+                if data.len() != 0 {
+                    return Err(Error::TrailingData);
+                }
+            }
+            None => return Err(Error::IncompleteMessage),
+        }
+
+        let mut substrings = input.splitn(2, ' ');
```

Comment:
Does `split_once` also work here? Is is better to introduce them to a new API, or to re-use the same API they used for handling the newlines? I don't know!

Reply:
We agreed not to change this